### DISCUSSION TITLE
Nats Consumer Log Message

### DISCUSSION THREAD

**Discussion by alberk8** [nats-io/nats.net#516] [Q&A] [answered]
I have a Jetstream consumer with the following log.  What does it mean by expires 00:00:30 and idle 00:00:15?  I want the consume to run as long as the aspnet app is running.  Do I need to worry about the message.  Thanks.

```
NATS.Client.JetStream.Internal.NatsJSConsume: Consume setup maxMsgs:1000, thresholdMsgs:500, maxBytes:0, thresholdBytes:0, expires:00:00:30, idle:00:00:15, hbTimeout:30000

[15:33:42.717  DBG][] NATS.Client.JetStream.Internal.NatsJSConsume: Sending pull request for init 1000, 0
[15:34:12.724  DBG][] NATS.Client.JetStream.Internal.NatsJSConsume: Header pending messages current 1000
[15:34:12.728  DBG][] NATS.Client.JetStream.Internal.NatsJSConsume: Header pending messages ["1000"] 0
[15:34:12.730  DBG][] NATS.Client.JetStream.Internal.NatsJSConsume: Check pending messages 0, 1000
[15:34:12.732  DBG][] NATS.Client.JetStream.Internal.NatsJSConsume: Sending pull request for pull-loop(chk-msgs) 1000, 0
```

**Sample Code**
 ```
var consumerConfig = new ConsumerConfig(NatsStreamConstant.SpotCheckConsumerName)
 {
     FilterSubject = "Data.ObjectSubject,
     DeliverPolicy = ConsumerConfigDeliverPolicy.All,
     AckPolicy = ConsumerConfigAckPolicy.Explicit,
     AckWait = TimeSpan.FromSeconds(10)
 };
  

 var consumer = await _natsJSContext!.CreateOrUpdateConsumerAsync("DataStream", consumerConfig);

 while (!cancellationToken.IsCancellationRequested)
 {
     await consumer.RefreshAsync(cancellationToken);

     await foreach (NatsJSMsg<Data<object>> msg in 
                       consumer.ConsumeAsync<Data<object>>().WithCancellation(cancellationToken))
     {
        //process data
     }
 }
```

**Comment by mtmk** [comment, marked as answer]
Those are settings for pull requests. Consume works with asking the server for messages and server sends back  as many as maxMsgs or if there aren't enough messages the pull request is closed by the server after 30 seconds (set by the expires setting). If there are no messages to be sent, server will send a heartbeat message every 15 seconds (idle setting). If the client receives all the messages it asked for, after the request is expired or doesn't receive any messages or heartbeat for more than 30 seconds (2 x idle) then it issues another pull request. the cycle continues until client terminates the consume loop. Bit of an explanation. please feel free to ask if I can clarify anything that was a bit confusing. cheers

**Reply by alberk8** [reply]
Thanks for the clarification.
